lib/srv/db/postgres/sql/update-permissions.sql

```diff
@@ -3,34 +3,97 @@ LANGUAGE plpgsql
+                    (obj.relkind = ANY (ARRAY['r', 'v', 'f', 'p']))
+                    -- only privileges we support provisioning.
+                    AND (acl.privilege_type = ANY (ARRAY['DELETE'::text, 'INSERT'::text, 'REFERENCES'::text, 'SELECT'::text, 'TRUNCATE'::text, 'TRIGGER'::text, 'UPDATE'::text]))
+                    -- only the user we are checking permissions for.
+                    AND grantee.rolname = username
+            ),
+            removed AS (
+                SELECT * FROM cur_perms
+                EXCEPT
+                SELECT * FROM new_perms
+            ),
+            added AS (
+                SELECT * FROM new_perms
+                EXCEPT
+                SELECT * FROM cur_perms
+            )
+        SELECT
+            (SELECT COUNT(*) FROM removed),
+            (SELECT COUNT(*) FROM added)
+        INTO
+            diff_removed,
+            diff_added;
 
-        SELECT COUNT(*) INTO diff_count_1 FROM (SELECT * FROM cur_perms EXCEPT SELECT * FROM new_perms) q1;
-        SELECT COUNT(*) INTO diff_count_2 FROM (SELECT * FROM new_perms EXCEPT SELECT * FROM cur_perms) q2;
-
-        IF (diff_count_1 > 0) OR (diff_count_2 > 0) THEN
-            RAISE WARNING 'Permission changes: removed=%, added=%', diff_count_1, diff_count_2;
+        IF (diff_removed > 0) OR (diff_added > 0) THEN
+            RAISE WARNING 'Permission changes: removed=%, added=%', diff_removed, diff_added;
             RAISE EXCEPTION SQLSTATE 'TP005' USING MESSAGE = 'TP005: User has active connections and permissions have changed';
         END IF;
     ELSE
         CALL pg_temp.teleport_remove_permissions(username);
 
-        -- Assign all roles to the created/activated user if grants are provided.
-        IF grant_data <> 'null'::JSONB THEN
-            FOR grant_item IN SELECT * FROM jsonb_array_elements(grant_data)
-            LOOP
-                EXECUTE 'GRANT ' || text(grant_item->>'privilege') || ' ON TABLE ' || QUOTE_IDENT(grant_item->>'schema') || '.' || QUOTE_IDENT(grant_item->>'table') || ' TO ' || QUOTE_IDENT(username);
-            END LOOP;
-        END IF;
+        -- Assign all privileges to the created/activated user if grants are provided.
+        FOR row_data IN SELECT * FROM new_perms
+        LOOP
+            EXECUTE format('GRANT %s ON %I.%I TO %I',
+                        row_data.privilege_type,
+                        row_data.table_schema,
+                        row_data.table_name,
+                        username
+                    );
+        END LOOP;
+
+        -- Grant USAGE on any schemas for which we made table privilege grants
+        -- See: https://github.com/gravitational/teleport/issues/51851
+        FOR row_data IN (
+            SELECT DISTINCT
+                table_schema
+            FROM
+                new_perms
+            WHERE
+                -- exclude system schemas
+                table_schema !~ '^pg_'
```

Comment:
what happens when admin attempts to grant USAGE on these?

is it safe to match `^pg_`? e.g. can an user create a schema named `pg_my_schema`